core/src/main/java/hudson/model/Node.java

```diff
@@ -374,15 +374,22 @@ public CauseOfBlockage canTake(Queue.BuildableItem item) {
         if(l!=null && !l.contains(this))
             return CauseOfBlockage.fromMessage(Messages._Node_LabelMissing(getDisplayName(), l));   // the task needs to be executed on label that this node doesn't have.
 
-        if(l==null && getMode()== Mode.EXCLUSIVE) {
-            // flyweight tasks need to get executed somewhere, if every node
-            if (!(item.task instanceof Queue.FlyweightTask && (
-                    this instanceof Jenkins
-                            || Jenkins.getInstance().getNumExecutors() < 1
-                            || Jenkins.getInstance().getMode() == Mode.EXCLUSIVE)
-            )) {
-                return CauseOfBlockage.fromMessage(Messages._Node_BecauseNodeIsReserved(getDisplayName()));   // this node is reserved for tasks that are tied to it
+        if(l==null && getMode() == Mode.EXCLUSIVE) {
```

Comment:
Ok so here, we first check whether the label is null and the mode is exclusive. If this is true, we will execute everything here. Otherwise, we'll instead execute the return CauseOfBlockage indicating that this node can't accept the task. Note that we've already checked that the label must exist and match the label for the node or else it would have exited already. This line did change but actually we probably shouldn't change it since there is no real change (just a whitespace change). I will remove that change and re-push the commit.